libvast/src/system/connector.cpp

```diff
@@ -0,0 +1,178 @@
+    case caf::sec::stream_init_failed:
+    case caf::sec::invalid_stream_state:
+    case caf::sec::unhandled_stream_error:
+    case caf::sec::bad_function_call:
+    case caf::sec::feature_disabled:
+    case caf::sec::cannot_open_file:
+    case caf::sec::socket_invalid:
+    case caf::sec::socket_disconnected:
+    case caf::sec::socket_operation_failed:
+    case caf::sec::unavailable_or_would_block:
+    case caf::sec::malformed_basp_message:
+    case caf::sec::serializing_basp_payload_failed:
+    case caf::sec::redundant_connection:
+    case caf::sec::remote_lookup_failed:
+    case caf::sec::no_tracing_context:
+    case caf::sec::all_requests_failed:
+    case caf::sec::field_invariant_check_failed:
+    case caf::sec::field_value_synchronization_failed:
+    case caf::sec::invalid_field_type:
+    case caf::sec::unsafe_type:
+    case caf::sec::save_callback_failed:
+    case caf::sec::load_callback_failed:
+    case caf::sec::conversion_failed:
+    case caf::sec::connection_closed:
+    case caf::sec::type_clash:
+    case caf::sec::unsupported_operation:
+    case caf::sec::no_such_key:
+    case caf::sec::broken_promise:
+    case caf::sec::connection_timeout:
+    case caf::sec::action_reschedule_failed:
+      return true;
+    case caf::sec::incompatible_versions:
+    case caf::sec::incompatible_application_ids:
+      return false;
+  }
+  return false;
+}
+
+bool is_recoverable_error(const caf::error& err) {
+  if (err.category() != caf::type_id_v<caf::sec>)
+    return true;
+  const auto err_code = std::underlying_type_t<caf::sec>{err.code()};
+  auto err_enum = caf::sec{caf::sec::none};
+  if (!caf::from_integer(err_code, err_enum)) {
+    VAST_WARN("unable to retrieve error code for a remote node connection "
+              "error:{}",
+              err);
+    return true;
+  }
+  return is_recoverable_error_enum(err_enum);
+}
+
+bool should_retry(const caf::error& err,
+                  std::size_t remaining_connection_attempts) {
+  return remaining_connection_attempts > 0 && is_recoverable_error(err);
+}
+
+} // namespace
+
+connector_actor::behavior_type
+connector(connector_actor::pointer self, std::uint64_t max_connection_attempts,
+          caf::timespan connection_retry_delay, vast::port port,
+          std::string host, caf::timespan connection_timeout) {
+  auto middleman = self->system().has_openssl_manager()
+                     ? self->system().openssl_manager().actor_handle()
+                     : self->system().middleman().actor_handle();
+  return {
+    [self, max_connection_attempts,
+     host](atom::connect) -> caf::result<node_actor> {
+      VAST_INFO("client connects to VAST node at {}", host);
+      return self->delegate(static_cast<connector_actor>(self),
+                            atom::internal_v, atom::connect_v,
+                            max_connection_attempts);
+    },
+    [self, port, host, middleman, connection_timeout, connection_retry_delay](
```

Comment:
The middleman should be captured in the state instead. Storing strong actor handles in a behavior can introduce cyclic references. It didn't in this case, but let's stick to best practices with CAF.